azurelinuxagent/ga/cgroupapi.py

```diff
@@ -470,55 +476,227 @@ def _get_root_cgroup_path():
+                continue
+
+            if expected_relative_path is not None:
+                expected_path = ""
+                controller_mountpoint = self._controller_mountpoints[controller]
+                if controller_mountpoint is not None:
+                    expected_path = os.path.join(controller_mountpoint, expected_relative_path)
+                if controller_path != expected_path:
+                    log_cgroup_warning("The {0} controller is not mounted at the expected path for the {1} cgroup; will not track metrics. Actual cgroup path:[{2}] Expected:[{3}]".format(controller, self._cgroup_name, controller_path, expected_path), send_event=False)
+                    continue
+
+            if controller == self.CPU_CONTROLLER:
+                controller_metrics = CpuMetrics(self._cgroup_name, controller_path)
+            elif controller == self.MEMORY_CONTROLLER:
+                controller_metrics = MemoryMetrics(self._cgroup_name, controller_path)
+
+            if controller_metrics is not None:
+                msg = "{0} metrics for cgroup: {1}".format(controller, controller_metrics)
+                log_cgroup_info(msg, send_event=False)
+                metrics.append(controller_metrics)
+
+        return metrics
+
+    def get_processes(self):
+        pids = set()
+        for cgroup_path in self._controller_paths.values():
+            with open(os.path.join(cgroup_path, "cgroup.procs"), "r") as cgroup_procs:
+                for pid in cgroup_procs.read().split():
+                    pids.add(int(pid))
+        return list(pids)
+
+
+class CgroupV2(Cgroup):
+    CPU_CONTROLLER = "cpu"
+
+    def __init__(self, cgroup_name, root_cgroup_path, cgroup_path, enabled_controllers):
+        """
+        :param cgroup_name: The name of the cgroup. Used for logging/tracking purposes.
+        :param root_cgroup_path: A string representing the root cgroup path. String can be empty.
+        :param cgroup_path: A string representing the absolute cgroup path. String can be empty.
+        :param enabled_controllers: A list of strings representing the agent supported controllers enabled at the root cgroup.
+        """
+        super(CgroupV2, self).__init__(cgroup_name)
+        self._root_cgroup_path = root_cgroup_path
+        self._cgroup_path = cgroup_path
+        self._enabled_controllers = enabled_controllers
+
+    @staticmethod
+    def get_supported_controllers():
+        return [CgroupV2.CPU_CONTROLLER, CgroupV2.MEMORY_CONTROLLER]
+
+    def check_in_expected_slice(self, expected_slice):
+        if expected_slice not in self._cgroup_path:
+            log_cgroup_warning("The {0} cgroup is not in the expected slice. Expected slice: {1}. Actual cgroup path: {2}".format(self._cgroup_name, expected_slice, self._cgroup_path), send_event=False)
+            return False
+
+        return True
+
+    def get_controller_metrics(self, expected_relative_path=None):
```

Comment:
Next step for cgroup v2 support will be adding logic to track metrics for cpu and memory in v2